server/api/rule.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"bytes"
+	"encoding/hex"
+	"net/http"
+	"strconv"
+
+	"github.com/gorilla/mux"
+	"github.com/pingcap/pd/pkg/apiutil"
+	"github.com/pingcap/pd/pkg/codec"
+	"github.com/pingcap/pd/server"
+	"github.com/pingcap/pd/server/core"
+	"github.com/pingcap/pd/server/schedule/placement"
+	"github.com/pkg/errors"
+	"github.com/unrolled/render"
+)
+
+var placementDisabledErr = errors.New("placement rules feature is disabled")
+
+type ruleHandler struct {
+	svr *server.Server
+	rd  *render.Render
+}
+
+func newRulesHandler(svr *server.Server, rd *render.Render) *ruleHandler {
+	return &ruleHandler{
+		svr: svr,
+		rd:  rd,
+	}
+}
+
+func (h *ruleHandler) GetAll(w http.ResponseWriter, r *http.Request) {
+	cluster := h.svr.GetRaftCluster()
+	if cluster == nil {
+		h.rd.JSON(w, http.StatusInternalServerError, server.ErrNotBootstrapped.Error())
+		return
+	}
+	if !cluster.IsPlacementRulesEnabled() {
+		h.rd.JSON(w, http.StatusPreconditionFailed, placementDisabledErr.Error())
+		return
+	}
+	rules := cluster.GetRuleManager().GetAllRules()
+	h.rd.JSON(w, http.StatusOK, rules)
+}
+
+func (h *ruleHandler) GetAllByGroup(w http.ResponseWriter, r *http.Request) {
+	cluster := h.svr.GetRaftCluster()
+	if cluster == nil {
+		h.rd.JSON(w, http.StatusInternalServerError, server.ErrNotBootstrapped.Error())
+		return
+	}
+	if !cluster.IsPlacementRulesEnabled() {
+		h.rd.JSON(w, http.StatusPreconditionFailed, placementDisabledErr.Error())
+		return
+	}
```

Comment:
It is really verbose of these lines of code. Maybe you want to try the middleware https://github.com/gorilla/mux#middleware

Reply:
I'll see what I can do for now.